### DISCUSSION TITLE
Symfony 7 not getting parameters from .env file

### DISCUSSION THREAD

**Discussion by azamalvi** [symfony/symfony#54415] [Q&A]
### Symfony version(s) affected

7.0.5

### Description

I have setup new project with Symfony 7. Now facing an issue while getting the value from .env (.env.local)
I tried both the `ParameterBagInterface` and direct in the controller.
So when I want to get the `APP_ENV` parameter value from .env using this code `$this->getParameter('APP_ENV')` it gives me this error

**You have requested a non-existent parameter "APP_ENV".**

But when I use this code `$_ENV['APP_ENV']` it works and this one also works `$this->getParameter('kernel.environment')`

My application is running in Dev mode but still I cleared the cache for both env and test but same error.

### How to reproduce

Create a new project using this command `symfony new my_project_directory --version="7.0.*"`
Then create a Controller and any Route then add this code in that route `dd($this->getParameter('APP_ENV'));` and you will get this error.

**Comment by MatTheCat** [comment]
[Environment variables](https://symfony.com/doc/current/configuration.html#configuration-based-on-environment-variables) and [parameters](https://symfony.com/doc/current/configuration.html#configuration-parameters) are two different things, so you cannot get an environment variable using `getParameter`.

Note that the `Autowire` attribute allows to inject parameters **or** environment variables, given the appropriate syntax: https://symfony.com/doc/current/service_container/autowiring.html#fixing-non-autowireable-arguments

**Reply by azamalvi** [reply]
@MatTheCat thanks for clearing. Can you please tell me what will be the best practice to get env variables? Should we define variables inside the `services.yaml` file and under `parameters` section by getting the value from .env file like this

```
parameters:
    app.env: '%env(APP_ENV)%'
```

and then we can get it `$this->getParameter('app.env')`?
My concern is that I just want to follow best practice to get different variables from .env.local file.
Thanks

**Reply by MatTheCat** [reply]
Best practices generally depend on use-cases, so I’m missing yours to answer: “get env variables” is way to vague.

The only thing I can say is that creating a parameter from an environment variable seems unnecessarily complicated.

**Reply by alexandre-le-borgne** [reply]
https://symfony.com/doc/current/configuration/env_var_processors.html

I think it's good practice to reference environment variables in parameters because it also allows you to cast/transform their value or set default ones.

**Reply by azamalvi** [reply]
Actually In my application there are bunch or variables that will be in .env.local file like "url, key, secret , etc" and it varies from server to server and i need to access these .env variables inside some services.

**Reply by MatTheCat** [reply]
Well this is still vague :sweat_smile: so I guess you should use [the `Autowire` attribute](https://symfony.com/doc/current/service_container/autowiring.html#fixing-non-autowireable-arguments): it’s simple and allow using processors.

Also, you should dump the local env vars in a PHP array when deploying: https://symfony.com/doc/current/configuration.html#configuring-environment-variables-in-production

**Reply by azamalvi** [reply]
In my previous Symfony application, I utilized a parameters.yml file alongside the parameter bag for configuration. However, in my new Symfony application, I've transitioned to using a .env file. While exploring solutions on [Stack Overflow](https://stackoverflow.com/questions/52151783/symfony-4-get-env-parameter-from-a-controller-is-it-possible-and-how), I noticed that many developers adopt a similar approach to mine: defining variables in services.yaml and accessing them via the parameter bag. I'm concerned about potential performance implications with this method.
